### ISSUE TITLE
句子的困惑度/得分计算？

### ISSUE THREAD

**Issue by xinfeng1i** [Morizeyao/GPT2-Chinese#66]
GPT2本质是上一个语言模型，那么对于训练好的Model，给定一段文本，应该可以计算这段文本的得分或者困惑度（Perplexity），这也是传统语言模型的应用场景之一。是否考虑增加一个脚本来支持该功能？

**Comment by lioyou** [comment]
你好，请问你已经实现了这个功能吗？
是这样做的吗？
output = model(prev, past=past)，
output[1] =  **prediction_scores**
**prediction_scores**: ``torch.FloatTensor`` of shape ``(batch_size, sequence_length,config.vocab_size)``
Prediction scores of the language modeling head (scores for each vocabulary token before SoftMax).
1 产生每个词时，保存对应词的一个概率
2 根据这个【https://blog.csdn.net/index20001/article/details/78884646】博文提供的函数计算值吗？

**Comment by fengzuo97** [comment]
困惑度可以通过np.exp(loss)来得到，loss可以通过model(input_ids, labels=input_ids)得到，label和输入一样。

**Comment by lioyou** [comment]
嗯，np.exe(loss)就是上面那篇博文的第二条等同公式。这样的确方便，另外，label与Input的确是一样的，计算时误差时pytorch-transformer做了位置切换，所以这个库训练时，实际没有必要区分label与input，重复一样的东西，浪费空间，input与label索引数组都是只读的，最后变成了Tensor。
谢谢！

**Comment by Morizeyao** [comment]
你说的没错，因为一开始我在我的代码里做了错位，当时是需要做区分的。后来发现没有这个必要，就改了回来，但是遗留了一个优化疏漏。

**Comment by xinfeng1i** [comment]
@lioyou 我参考 XLNet 计算困惑度方式，实现了一个版本，它的基本思路和你们前面讨论的一样：直接计算 `math.exp(loss)`  即可。代码如下，我测试了一下，结果是符合预期 ^_^

```
@torch.no_grad()
def get_score(sentence):
    tokens = tokenizer.tokenize(sentence)
    input_ids = tokenizer.convert_tokens_to_ids(tokens)
    input_tensor = torch.tensor([input_ids])
    outputs = model(input_tensor, labels=input_tensor)
    loss, logits = outputs[0:2]
    return math.exp(loss.item())
```

**Comment by xinfeng1i** [comment]
语言模型使用的交叉熵损失，所以
    

- 第一个词语的损失为 -log p(w_1)

    

- 第二个词语的损失为 -log p(w_2|w_1)

  

-   ...

    

- 第 n 个词语的损失为 -log p(w_n|w_1...w_n-1)

Pytorch的[交叉熵损失](https://pytorch.org/docs/stable/nn.html?highlight=crossentropyloss#torch.nn.CrossEntropyLoss)函数默认会取平均，所以总的Loss为

```
Loss = {[-log p(w_1)] + [-log p(w_2|w_1)] + ... + [-log p(w_n|w_1...w_n-1)]} / n
        = {-log p(w_1) * p(w_2|w_1) * p(w_n|w_1...w_n-1)} / n
        = {-log p(w_1 w_2 ... w_n)} / n
```

而[困惑度的定义](https://towardsdatascience.com/perplexity-intuition-and-derivation-105dd481c8f3)为： PP = 2^{-L}，其中 L = { log P(w1 w2 ... w_N) } / N

上面的 `-L` 等于 `Loss`，所以直接对 Loss 取以 2 为底的指数即可，然而以 2 对为底还是以 e 为底没有本质区别。所以直接用 `exp(loss)` 计算从数学上来讲是完全正确的 ^_^

**Comment by lioyou** [comment]
我就知道您已经实现了这个功能，不过利用exp(loss)，前提是交叉熵计算误差，我就不Pull了，要不您来，当然前提是作者觉得这个Pull有意义，但是这个库本身不具备扩展性(硬编码)，加得越多，显得越乱，相信大家都有自己重构的版本。
谢谢！

**Comment by Morizeyao** [comment]
对的，之前接受了几个pull，导致代码有点乱，实际上我写的eval.py里已经有用math.exp(loss)计算ppl的功能了。

**Comment by lioyou** [comment]
作为一个开源项目，作者你真的觉得是大家的Pull导致你代码乱吗？如果你自己真的运行过代码，就不会连参数名称错误，类型错误都不知道吧，另外无意义的代码，逻辑错误等在Stable版本依旧在，这你自己写的吧？当然我自己也犯过，但我承认。另外，你有时间看一下这些Pull，更多的是修复错误，使代码条理更加清晰。而现在这个master库还存在一些逻辑错误，参数一改肯定有问题。
唉，我还是说声谢谢吧，毕竟这个项目只要别动还能跑，以这个项目为起点，仔细阅读每行代码的意义，我也学到更多，与大家的交流，更让我受益匪浅。最后也祝这个项目越来越好吧。
谢谢！

**Comment by Morizeyao** [comment]
行吧。

**Comment by xinfeng1i** [comment]
实现比较简单，我觉得可以不用提交 PR 了，后续有这个需求的同学可以参考这个答案。这个 Iusse 可以关闭了。
